MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,1771 @@
+  //   -   In the double loop: assign a new fitted vertex to the one resolving shared jets
+  //   -   loop thru a set of vertices after the mitigation to clean up a vertex with just one track
+  // Note that:
+  //   - {1,1} shared jets have exactly one track to the jet from both vertices
+  //   - {1,n} shared jets have one of the two vertices contributing exactly one track to the jet
+  //////////////////////////////////////////////////////////////////////
+  if (resolve_shared_jets) {
+    edm::Handle<pat::JetCollection> jets;
+    event.getByToken(shared_jet_token, jets);
+
+    std::vector<std::vector<size_t> > sv_total_track_which_trkidx;
+    std::vector<unsigned int> sv_total_ntrack;
+    std::vector<unsigned int> sv_ascending_total_ntrack;
+    std::vector<size_t> sv_ascending_vtxidx;
+
+    std::vector<std::vector<int> > sv_match_tracktojet_which_trkidx;
+    std::vector<std::vector<int> > sv_match_tracktojet_which_jetidx;
+
+    std::vector<track_vec> sv_total_track_which_trk_vec;
+
+    int n_output_aftersharedjets_onetracks = 0;
+
+    size_t vtxidx = 0;
+    for (v[0] = vertices->begin(); v[0] != vertices->end(); ++v[0]) {
+      std::vector<size_t> track_idx;
+      std::vector<int> tracktojet_which_trkidx;
+      std::vector<int> tracktojet_which_jetidx;
+      track_vec tks = vertex_track_vec(*v[0]);
+      sv_total_track_which_trk_vec.push_back(tks);
+      for (size_t i = 0; i < tks.size(); ++i) {
+        const reco::TrackRef& itk = tks[i];
+        track_idx.push_back(i);
+        for (size_t j = 0; j < jets->size(); ++j) {
+          if (match_track_jet(*itk, (*jets)[j], *jets, j)) {
+            tracktojet_which_trkidx.push_back(i);
+            tracktojet_which_jetidx.push_back(j);
+            if (verbose)
+              printf(" track %u matched with a jet %lu \n", tks[i].key(), j);
+          }
+        }
+      }
+
+      unsigned int ntracks = track_idx.size();
+
+      if (vtxidx == 0) {
+        sv_ascending_total_ntrack.push_back(ntracks);
+        sv_ascending_vtxidx.push_back(vtxidx);
+      }
+      else {
+
+        std::vector<unsigned int>::iterator it_ntracks = sv_ascending_total_ntrack.end();
+        std::vector<size_t>::iterator it_vtx = sv_ascending_vtxidx.end();
+        while (it_ntracks != sv_ascending_total_ntrack.begin() && ntracks <= sv_ascending_total_ntrack[std::distance(sv_ascending_total_ntrack.begin(), it_ntracks)-1])
+        {
+          --it_ntracks;
+          --it_vtx;
+        }
+
+        if (it_ntracks == sv_ascending_total_ntrack.end() && ntracks > sv_ascending_total_ntrack[std::distance(sv_ascending_total_ntrack.begin(), it_ntracks)]) {
+          sv_ascending_total_ntrack.push_back(ntracks);
+          sv_ascending_vtxidx.push_back(vtxidx);
+        }
+
+        else {
+          sv_ascending_total_ntrack.insert(it_ntracks, ntracks);
+          sv_ascending_vtxidx.insert(it_vtx, vtxidx);
+        }
+      }

```

Comment:
Comments in this section would be useful since it isn't immediately obvious why the `vtxidx == 0` case is different, or why `push_back` vs. `insert` are used in the `if/else` towards the bottom of what is highlighted, rather than being able to just stick with one of them (I am guessing it must be related to the order in which you iterate through the tracks or something).

Reply:
Oh, maybe this is to sort the vertices by ntracks? Either way, comments will help.

Reply:
Done with more understandable comments.